Implement LoggingAsync for GRPC Worker Services

xldrx opened the pull request:
This allow RecvTensor events to show up in StepStats and in turn in Chrome Tracing format.

xldrx committed:
Implement LoggingAsync for GRPC Worker.

xldrx committed:
Add nullptr checks

xldrx committed:
Fix BUILD file format

tensorflow-jenkins commented:
Can one of the admins verify this patch?

xldrx commented:
@pbar @mrry @suharshs Could you please take a look at this?

mrry commented:
Reassigning this to @prb12, since he knows the ins and outs of the logging path better than I do. A few quick observations though:

* Instead of modifying the `BaseRendezvousMgr` and related classes to capture this information, please use the existing stubs in `GrpcWorkerCache`:

  https://github.com/tensorflow/tensorflow/blob/c0662f1620c2b97abb79b8ae6a8a30f7c7719475/tensorflow/core/distributed_runtime/rpc/grpc_worker_cache.cc#L86-L92

* The changes in `master_session.cc` are potentially controversial, because they could cause a lot of additional logging to be generated when `FULL_TRACE` is specified. We might need an additional `TraceLevel` for this case.

* Run `clang-format -style=Google` over the code to take care of style issues (which should trigger a failure in the presubmit).

xldrx committed:
Change LoggingAsync implementation

- Revert changes to *_rendezvous_mgr
- Implement logging primitives in session_mgr instead
- Implement ClearLogs
- Fixed C++ formating

xldrx commented:
Thanks @mrry for the feed back:
* I changed the implementation to use SessionMgr instead of *RendezvousMgr. 
Note that unlike other worker rpcs, `LoggingRequest` in  `LoggingAsync` rpc does not have any session handle, therefore all sessions should be touched. Both implementations (previous one, and this one) iterates over all sessions (previously through `RendezvousMgr` now `SessionMgr`), then uses `GrpcWorkerCache` calls in each session.

* @pbar: Thoughts on changes in `master_session.cc`?

xldrx committed:
Check for nullptr

sb2nov commented:
@xldrx can you look into the merge conflicts.

xldrx committed:
Merge branch 'master' into master

xldrx commented:
Hey @sb2nov,
It seems that Paul is swamped. Could you review this PR?

sb2nov commented:
@xldrx I don't know enough to review this right now. @prb12 ping.

prb12 commented on tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc:
```diff
@@ -392,6 +392,24 @@ void GrpcWorker::GrpcRecvTensorAsync(CallOptions* opts,
+      session_mgr->SetLogging(request->rpc_logging());
+      if (request->clear()) {
+        session_mgr->ClearLogs();
+      }
```
If both `clear` and `fetch_step_id` are used the semantics are unclear, but I would have thought it made more sense to retrieve any previously collected data before clearing?

xldrx replied:
Great observation. Fixed.

prb12 reviewed (changes requested):
Thanks for addiing this!

With regard to an additional TRACE_LEVEL - I don't think that this is likely to be a problem.  

I'm assuming this doesn't have a noticeable performance impact when tracing is not enabled?

One minor question about the case where clear and retrieve are done at the same time.

xldrx committed:
Merge https://github.com/tensorflow/tensorflow

xldrx committed:
Better handling of the case when both "clear" and "retrieve" flags are sent to "LogAsync."
nullptr check on default_worker_cache_.

xldrx committed:
Fix formatting

xldrx commented:
@prb12 Thanks for the review. 
> I'm assuming this doesn't have a noticeable performance impact when tracing is not enabled?

This PR does nothing when tracing is not enabled.

> One minor question about the case where clear and retrieve are done at the same time.

Great observation. Fixed.

rmlarsen commented:
@xldrx This appears to have numerous broken tests. Can you please take a look?

xldrx committed:
Merge https://github.com/tensorflow/tensorflow

xldrx committed:
Updata session_mgr.cc to address changes in 619792f

rmlarsen commented:
@xldrx Thanks for the contribution!
